sycl/include/CL/sycl/detail/program_impl.hpp

```diff
@@ -340,9 +341,11 @@ class program_impl {
     }
   }
 
-  void create_cl_program_with_il(OSModuleHandle M) {
+  void create_pi_program_with_il(OSModuleHandle M,
+                                 const string_class &KernelName) {
     assert(!Program && "This program already has an encapsulated PI program");
-    Program = ProgramManager::getInstance().createOpenCLProgram(M, Context);
+    Program =
+        ProgramManager::getInstance().createPIProgram(M, Context, KernelName);
```

Comment:
Kernel resolution has been decoupled from PI API and ProgramManager::loadProgram method in the current version of the patch, but I left it in createPIProgram and getBuiltPIProgram public methods for now. Since the kernel resolution map is populated and maintained at registration time, access to it has to be guarded with the global lock the same way access to the device image map is. Correct me if I'm wrong, but if the lock was to be released between kernel resolution and program creation/building, the result of the former might be invalidated by another thread, so providing a public method that does both underneath makes sense to me. Let me know your thoughts on this.

Reply:
@kbobrovs ping